### DISCUSSION TITLE
Poll: How will you use Home Assistant with AmpliPi?

### DISCUSSION THREAD

**Discussion by linknum23** [micro-nova/AmpliPi#206] [Ideas]
We know many of you are planning on integrating AmpliPi into your Home Assistant (HA) setups.

One developer has already started working on a prototype HA integration, see #184 for the initial plan, but we would like to use their resources efficiently.

What HA integrations and automations would you create, assuming AmpliPI was available in HA?
Please add an answer that describes the functionality you would like or up-vote others.

**Comment by bs42** [comment]
Home Assistant has a handler for media device groups, as an example, both the Sonos and Yamaha Music Cast integrations support them.  IMO exposing each zone as a device of sorts to HA will fit in with the existing method of use that already exists in HA.

I don't see any real benefit of exposing the AmpliPi's preset groups into HA, but that could just be me.  HA can have it's own control over speaker groups and to me, it makes the most sense for it to do that on it's own.

With the current media players such as Sonos and Music Cast (I use these as examples because I have used them both), there are Sources which are the various physical and digital sources, i.e. Amazon Music, Line In, etc.  Having the AmpliPi populate the sources for each of the zones when exposed to HA would provide the most seamless experience for existing HA users.

For me, it's primarily the adding/removing of zones from a specific source that I use constantly.

Examples:  

The garage is playing the "main source" of audio when there is "presence" in the garage.

The Family room, and kitchen are playing the "main source" all the time, except when the TV is turned on at which point that zone is turned off so that the audio source is now the TV instead.

Backyard speakers are joined to the "main source" when the "backyard activity" input is enabled in HA.

Planned usage is for the kids rooms to each have the ability to stream their own bluetooth devices into the speakers in their rooms with the ability for HA to "play 'dinner is ready' flac file and then resume their bluetooth".

**Reply by linknum23** [reply]
Thanks for the examples and feedback, this is really useful.

**Comment by Juggler00** [comment]
I would agree with most of what @bs42 has written. I do want to share my current setup that the AmpliPi will be replacing.

I currently have 5x Chromecast Audios, each attached to a separate amplifier. Through the Google Home app, I'm able to pre-define "zones" (i.e. combos of speakers). This gives me the ability to ask my Google Home to play local internet radio in the kitchen, or a specific Spotify album on "inside". Homeassistant ties in with this through wall-mounted tablets that gives me the ability to control local volume as well as play/pause/skip tracks. 

In an ideal world, I would still be able to use some sort of voice assistant to get the AmpliPi to play a local internet radio station or Spotify resource on a specific set of AmpliPi outputs. I would also be able to do this through a wall-mounted tablet running Lovelace.

**Comment by marc-portier** [comment]
just stepping into the world of both HA and AmpliPi for the ongoing house-redo-project... so sorry if my input is naive or vague:

1.  I now hope HA + Niko Home Control will be actually  doing what I suggested in #203 -- have extra simple buttons to control volume + mute in some particular zone - without the custom hardware solution this makes those buttons more flexible
2.  still thinking about how a simple button could also advance me through preset-sources (although previous comments here make me feel one might want to let HA control that primarily and drive AmpliPi? 
3.  having HA on board is adding the "DOORBEL" idea to my wish-list --> hoping I can have a chosen wav played into selected zones (bot possibly controlled by knowledge inside HA , like time of day, or 'child sleeping' mode) when the button-at-the-door is pressed

**Comment by Saars1337** [comment]
I'm putting the new Sonoff NSPanels in each room

I'm not entirely sure how I would like it, but these scenarios seem nice to me
1. Have the current streams listed and allow users to toggle them for the specific room they are in
2. Be able to play a set playlist from something like Spotify and have it default to the room they are in
3. Having some kind of PA function for things like doorbell is critical for me, I want this system to be able to replace all my Google Home Mini's I have in each room. One of the many pains in the extremely complicated Node-Red runbook I have that saves the audio states, plays the doorbell sound  at max volume, and then goes back to whatever it was doing before

So having a control panel in Home Assistant that lists each source and a toggle for each zone so you can quickly turn them all on/off from a central panel

And pretty please don't make "Groups" a separate entity like Google do (e.g when I play something to "Home Group", even though is plays on all speakers, each speaker independently shows idle in HA instead of playing)

**Reply by linknum23** [reply]
We have been watching those Sonoff NSPanels, it looks like several developers are working on [changing the firmware](https://community.home-assistant.io/t/sonoff-nspanel-smart-scene-wall-switch-by-itead-coming-soon-on-kickstarter/332962/48) to interface directly with HA and use it as an HA control panel? Or are you planning on using Sonoff's interface?

The PA function is already integrated into AmpliPi (it handles all of the save/restore behind the scenes). Hopefully this can be exposed by the HA integration.

We will keep the rest of these ideas in mind as well.

Thanks for the feedback!

**Reply by bs42** [reply]
Thanks for the heads up about those Sonoff NSPanels, I had not seen that and am VERY interested in the progress of that firmware mod.

**Reply by Saars1337** [reply]
No chance I'm using them with the Sonoff firmware, I had hoped that someone would write custom HA firmware for it by the time they arrived, otherwise, I'll do it myself but I am definitely no expert

**Comment by rverstappen** [comment]
I'm also busy setting up a Home Assistant platform and I'm looking to replace an aging AMX Precis DSP 8x8 audio switcher with something like AmpliPi if it would have HA support. I think you need to do whatever possible to make AmpliPi discoverable and not require any configuration.yaml. This seems to be possible with other similar HA integrations. 

At the very least, you will need to implement the Media Player interface, presumably one for each output: https://www.home-assistant.io/integrations/media_player/

The Universal Media Player interface makes total sense too. https://www.home-assistant.io/integrations/universal/

It looks like Brian has the right idea to develop a standalone Python API module first (I see it in his development branch) and use that to implement the HA integration.

**Reply by linknum23** [reply]
His integration is in progress here: https://github.com/brianhealey/hacs_amplipi

**Comment by brianhealey** [comment]
The way I understand it, the AmpliPi has 4 DAC's that it is capable of streaming at a given time. You can see this modeled in the UI for the device and the developer API.

The HA integration path I chose was to create 4 `media_player` entities to model the DAC's. I originally created one for each zone, but quickly found that led to awkward behavior as changing the source on one zone would automatically change it on others. 

I don't have an actual device, so I have been developing against a mocked developer service running in a VM. I believe the mute/volume controls should be working and I am just smoothing out the play/pause/next controls for pandora/spotify.

Right now I don't support `group/join` and I haven't been able to test `tts`, changing sources, etc. Use at your own risk and please provide feedback as you play with it.

https://github.com/brianhealey/hacs_amplipi

**Reply by bs42** [reply]
For what it's worth, I'll throw in my 2 cents.

Here's what I'd like to see:
 
media_player entities for each DAC (media_player.amplipi_source_1,2,3,4) (aka amplipi sources) that has a sources list that is "local" and the full list of streams.

media_player entities for each zone (media_player.amplipi_zone_1,2..16) that has a sources list that are the amplipi sources

To me, this gives the best control, it's how I've been operating using the universal media player and REST commands (though not complete, I don't have the sources all working as source_list because I couldn't figure out how to mimic that in REST).  

The zones have the volume controls and can choose which "amplipi source" to play.  The sources can have their streams manged as well via HA.

**Reply by brianhealey** [reply]
Without having the hardware in hand, its hard for me to predict the behavior. 

I can see changing volumes "per zone" but when you change a source of a zone, you would be actually changing the source to the DAC that zone is associated with, which may not be expected behavior. If I was to suggest a middle ground, I will create the DAC `media_player` entities and set them `enabled` by default. I could also then create the `zone` `media_player` entities and default to `disabled`. That way the user can decide if they want them exposed or not. :)

**Reply by bs42** [reply]
I probably wasn't clear on my intent.

I agree that the zones should **not** have the ability to change a DAC's source.  I meant that the sources list for a zone's `media_player` should be the 4 DAC's.  And then the DAC's `media_player` entity would have it's sources list be all the streams+"input".
```
media_player.amplipi_input_1
- source_list:
   - Local
   - Spotify
   - Pandora
   - last.fm
```

```
media_player.amplipi_zone_1
- source_list:
   - Input 1
   - Input 2
   - Input 3
   - Input 4
```
Without the ability to control a zone in HA, there's no way to set volume or control which DAC a zone is on.  That is why I propose DACs and Zones all have their respective `media player`'s exposed to HA.

I hope that's a better explanation.

Thanks for your work on this, it's really appreciated.

**Reply by brianhealey** [reply]
Thanks for the explanation. I see what you are saying and that does sound like a useful way of using it.

There is a feature in `media_player` called `SUPPORT_GROUPING` might expose additional capabilities in HA. I think I can do what you are proposing for a phase1 and in the spirit of open source, we can all make it better later.

**Reply by Juggler00** [reply]
I've added your code through the HACS custom repositories. Thank you for all of your work thus far!

Would you prefer feedback here, or directly in your GitHub?

**Reply by brianhealey** [reply]
I think here is fine for now until we have a "release". It's still a wip progress, but I am curious to hear how it works in its current state.

**Reply by Juggler00** [reply]
Here is my setup:
- HA v2022.2.0) as a docker container.
- 3 additional inputs (beyond rca 1-4) configured on AmpliPi: Airplay, 2x streaming internet stations

On my Lovelace dashboard I see 4x media player entities, 1 for each of the inputs. When "idle", they each have a play button, media button (this does nothing), and the 3-dot more button. Clicking on the more button brings up the media pop-up that also has volume and TTS input. The TTS input works!

When streaming an internet station (initiated from AmpliPi web interface), the media player entity changes to reflect the artwork of the station. The play button becomes a pause button and there is also a next button. Pressing the pause button gives an error "Failed to call service media_player/media_pause. AmpliPi api error: The url http://192.168.2.15:80/api/streams/0/pause returned error 404". Clicking on the more button shows the same as before, note the volume slider does not reflect the current level set through the AmpliPi interface. 

When the input is changed to the Airplay stream, the media player entity shows the title "Airplay". When I "cast" to the input, the play also changes to pause and the next track button appears. 

The volume control as previously noted does not reflect the current volume level. Changing the volume through HA does not change the volume.

I fully appreciate that this is a beginning and am very grateful for your work. Happy to test at any point! Can I ask what your involvement with AmpliPi is? If you don't have a unit, why are you developing this?

**Reply by brianhealey** [reply]
Thanks for the feedback. I will put out another release soon that includes different logic for when the media controls show. I think they are reversed right now.

Love that the TTS works. That is one of my favorite use cases.. Coming in your front door and having HA welcome you home. :)

Interesting about the volume controls. Most likely this is do the the fact that I am developing against a feature branch that handles volume controls differently. @linknum23 Do you have any input on when this will be released?

I have an AmpliPi on order :).  I originally signed up for the kickstarter campaign, but backed out due to timing on my part. I have 4 raspberry Pi devices running DAC/AMP hats today, but have additional zones that I need to add. This device has everything I need sans HA integration, so I figured I could get a jump on that.

**Reply by Juggler00** [reply]
Fantastic. I will continue to monitor your GitHub and provide feedback as you develop. Amazing work!

**Reply by linknum23** [reply]
We will see what we can do to get a release out soon. First thing is to add some tests to #242 so it can be merged into devel. 

In the meantime here is an update file you can use to test just that branch:
[amplipi-0.1.8a0.tar.gz](https://github.com/micro-nova/AmpliPi/files/8017656/amplipi-0.1.8a0.tar.gz)

To update with that file:
1. Go to **Settngs**->**Updates** ->**Check for Updates**
2. Click on the **Custom** tab
3. Browse for amplipi-0.1.8a0.tar.gz
4. Click **Start Update**

**Reply by brianhealey** [reply]
Thanks! I have a physical device in hand now and applied that firmware. I can see the audio/mute controls are working against my latest code base. I will fixup the media player controls and look at adding the per-zone media players as well.

**Comment by linknum23** [comment]
Right now AmpliPi only supports playback control on Pandora and Spotify :/
We need to add more support to come of the other streams and add some way
of indicating each of the streams capabilities. Sorry about that.

**Comment by sepffuzzball** [comment]
Just throwing in some of the use cases I have for Home Assistant + AmpliPi:

- Motion Sensor audio: all of my rooms have some level of basic motion sensing, so in 'all speaker mode' I'd like to actually control zones by using these motion sensors (and contact sensors for things like bathrooms) so it's not actually playing everywhere, just where humans are
- Individual presence streams: I have Raspberry Pi W's in every room running room-assistant which can use bluetooth to determine who is occupying specific rooms (usually tied to a phone or smartwatch) and as an advanced function, I'd like to have on of the AmpliPi streams 'assigned' to a person and have it follow said person (such as if I'm listening to music in the upstairs office and my spouse is listening in the downstairs office, if I move into the kitchen have the upstairs office zone fade out of 'my' stream and attach the kitchen zone until I go back upstairs. I expect this to be _fun times_ in NodeRed
- Every year we do more and more sophisticated stuff for Halloween, whether it's hand-wave tripwire lidar sensors to play certain audio files (potentially in different outdoor/garage zones) or play sound effects when _a thing_ has occurred (currently prototyping an accelerometer cube tied into Home Assistant for interactivity) and doing _things_ when it's on the correct facing. Having multiple streams and zones and being able to assign them dynamically will make things even more excellent (previously was using multiple speaker sets and chromecast audios, but the chromecast audio can be slow to wake)
- Nabu Casa's TTS (tts.cloud_say) for semi-sophisticated door announcements using a Google Coral compute device and likely Frigate to have better doorbell detection (such as _Package delivery detected at the front door_ or _Welcome home <resident name>_ or greeting a guest or someone it doesn't recognize)

**Reply by brianhealey** [reply]
I have similar use cases (occupation sensing) and TTS announcements using a doorbell cam.  

Instead of Raspberry Pi's in each room, I am using cheap esp32 sensors running espHome. They are super easy to set up and they can  sense BLE devices as well as act as iBeacons for any client driven integrations. I also have an NSPanel on the way (from AliExpress so its a bit out there) that I plan on flashing with espHome and setting up a simple HA assistant frontend.

The AmpliPi HA integration will provide the `media_player` entity which supports TTS already. You will be able to automate muting and changing the source of an input.

**Reply by sepffuzzball** [reply]
Excellent! And thanks for getting started on the HA integration, I've actually got it installed in my instance (though I'm still on the shipping firmware of the AmpliPi) just haven't had a chance to play with it just yet!

Honestly I probably should just be using ESPHome for presence detection since I use it extensively for all my lighting/plugs/switches (I currently have 82 devices in ESPHome)

**Reply by Swamp-Ig** [reply]
With 82 devices on wifi, does your wifi get very slow?

**Reply by sepffuzzball** [reply]
Not really! They're all mostly on 2.4Ghz and most of our things like laptops and phones are on 5Ghz. I've got 4 access points (Unifi 6 LR/Pro depending on location) so it hasn't been a problem!

**Comment by brianhealey** [comment]
I posted a newer release last night that resolves the following:
- Mute controls now work with 0.1.7 releases against sources only
- Volume controls now work with 0.1.7 releases against sources only
- Resolved some issues where the artwork was not rendering properly
- Media controls now show play/pause, previous, and next (resolves some of the odd 404's)
- Device status now includes the firmware version
- Device status now includes a link to the AmpliPi
- Media player title will switch between the track name (if available) and the source name
- Media player "app" attribute is now set to the appropriate source type (spotify, plexamp, etc)

I will be adding the "zone" media players within the next few days. The distinction will be:
- Zones will be disabled by default but can be enabled by the other
- Zones do not allow you to change the source stream
- Changing the zone "stream" changes zone group membership instead
- Zones can be individually enabled/disabled (on/off)

**Reply by brianhealey** [reply]
Updated release tonight. 
- Resolved an issue where the input source could not be reselected (Thanks @hawkeyeP )
- Zones and Groups are now added asa media player entities (default to disabled)
- Zone/Groups can be turned on and off (enabled/disabled in the web ui but for me they keep playing...)
- Volume support for individual zones/groups
- Mute support for individual zones/groups

I am going to be playing with the HA grouping support for groups/zones to determine my next steps, but honestly it should be usable by most in its current form.

I will be working on documentation, hardening the code, and testing against development firmware releases.

**Reply by Juggler00** [reply]
Very nice release. Able to control volume, see artwork, select source for each input.

**Comment by brianhealey** [comment]
I apologize for the slow updates. I had some time to play with the zone/group assignment and it seems to work nicely.  

The feedback has been great, please keep it coming!

**Reply by bs42** [reply]
No apologizes necessary!

Thanks for the update to get it working in the new HA release.
